mnist/mnist_gcp.ipynb

```diff
@@ -0,0 +1,895 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# MNIST E2E on Kubeflow on GKE\n",
+    "\n",
+    "This example guides you through:\n",
+    "  \n",
+    "  1. Taking an example TensorFlow model and modifying it to support distributed training\n",
+    "  1. Serving the resulting model using TFServing\n",
+    "  1. Deploying and using a web-app that uses the model\n",
+    "  \n",
+    "## Requirements\n",
+    "\n",
+    "  * You must be running Kubeflow 1.0 on GKE with IAP\n",
+    " "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prepare model\n",
+    "\n",
+    "There is a delta between existing distributed mnist examples and what's needed to run well as a TFJob.\n",
+    "\n",
+    "Basically, we must:\n",
+    "\n",
+    "1. Add options in order to make the model configurable.\n",
+    "1. Use `tf.estimator.train_and_evaluate` to enable model exporting and serving.\n",
+    "1. Define serving signatures for model serving.\n",
+    "\n",
+    "The resulting model is [model.py](model.py)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Verify we have a GCP account\n",
+    "\n",
+    "* The cell below checks that this notebook was spawned with credentials to access GCP\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import os\n",
+    "import uuid\n",
+    "from importlib import reload\n",
+    "from oauth2client.client import GoogleCredentials\n",
+    "credentials = GoogleCredentials.get_application_default()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Install Required Libraries\n",
+    "\n",
+    "Import the libraries required to train this model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "pip installing fairing git+git://github.com/kubeflow/fairing.git@9b0d4ed4796ba349ac6067bbd802ff1d6454d015\n",
+      "Checkout kubeflow/tf-operator @9238906\n",
+      "Configure docker credentials\n",
+      "Adding /home/jovyan/git_tf-operator/sdk/python to python path\n"
+     ]
+    }
+   ],
+   "source": [
+    "import notebook_setup\n",
+    "reload(notebook_setup)\n",
+    "notebook_setup.notebook_setup()"
```

Comment:
I think a kernel restart is needed at this point.  Also, I needed to install 'retrying' too, do the kernel restart  (I added a cell with `!pip install --user retrying`)

Reply:
I'm like to try to find a way to avoid the kernel restart. Because I'd like to just be able to tell people to do "Run All."

I believe if we update the path the imports should be locatable immediately. If you've already loaded the module a reload might be needed.

Reply:
Hmm, it looked from the output like the paths were updated when I ran `notebook_setup()`, but I still had to restart.  Maybe we should see if someone else can repro. (@jinchihe , if you run through it too, let us know if this part worked for you).

Reply:
I tried the latest version in a new notebook
https://github.com/jlewi/examples/commit/5b1262fe6255776d03dbded7d4d500a04508491f

The imports worked fine.
I have added reload statements since the original verison.